Add missing minor security checks

julek-wolfssl opened the pull request:
- max number of encryption
 - max number of failed decryptions
 - max number of rekeying

julek-wolfssl committed:
Additional checks for secure renegotiation

julek-wolfssl committed:
Refactor nofail tests into test_wolfSSL_client_server_nofail

julek-wolfssl committed:
Enforce maximum amount of failed decryptions in DTLS 1.3

julek-wolfssl committed:
TLS 1.3: Check maximum records encrypted with one key set

rizlik commented on tests/api.c:
```diff
@@ -55306,6 +55306,148 @@ static int test_wolfSSL_dtls_fragments(void) {
+
+    test_wolfSSL_client_server_nofail(&func_cb_client, &func_cb_server);
+
+    AssertTrue(func_cb_client.return_code);
```
should we return -1 here

julek-wolfssl replied:
Fixing.

rizlik commented on src/dtls13.c:
```diff
@@ -1998,6 +1998,13 @@ int Dtls13NewEpoch(WOLFSSL* ssl, w64wrapper epochNumber, int side)
+    /* We are updating the receiving keys for this connection. We can restart
+     * the failed decryption counter. */
+    if (side == ENCRYPT_AND_DECRYPT_SIDE || side == DECRYPT_SIDE_ONLY)
+        w64Zero(&ssl->macDropCount);
```
I think this will not be precise, we have multiple epochs and we can change epochs, but `macDropCount` is shared. I think it would be more correct to maintain a per epoch counter.

rizlik replied:
Also, consider removing an epoch if, once we receive the ack of the key update, its macDrop counter is greater than our limit. We will not decrypt anymore with that epoch.

julek-wolfssl replied:
I don't think we need a per epoch granularity. Adding marking old epochs as encrypt only so that they can not be used for decryption.

rizlik replied:
But then you have also to mark old epochs as encrypt only every time you receive a new decrypting epoch, otherwise sending a new decrypting epoch can be used to reset the counter. I think adding a per-epoch counter may be a more explicit, clean, and easy-to-audit solution.
Is the memory usage concerning you? 
Another option would be not allowing multiple decryption epochs after the handshake, but it would be a major refactor and we still need mulitple epochs logic for the handshake.

julek-wolfssl replied:
Added a per epoch counter. My main concern was not changing too much with this PR.

rizlik commented on src/internal.c:
```diff
@@ -18818,6 +18818,9 @@ static int checkDTLS13AEADFailLimit(byte bulk_cipher_algorithm,
             break;
 #endif
+        case wolfssl_cipher_null:
+            /* No encryption being done. The MAC failed must have failed. */
```
is the comment missing a not?

julek-wolfssl replied:
Updating to `No encryption being done. The MAC check must have failed.`

rizlik commented on src/tls13.c:
```diff
@@ -2103,7 +2103,7 @@ static WC_INLINE void WriteSEQTls13(WOLFSSL* ssl, int verifyOrder, byte* out)
 #endif /* WOLFSSL_DTLS13 */
     }
-    else if (verifyOrder) {
+    else if (verifyOrder == PEER_ORDER) {
```
Is this an unrelated bug?

julek-wolfssl replied:
Just wanted to make this explicit. No bug found.

rizlik commented on src/tls13.c:
```diff
@@ -11134,7 +11134,7 @@ int Tls13UpdateKeys(WOLFSSL* ssl)
     /* TODO: add WOLFSSL_ERROR_ALREADY_IN_PROGRESS type of error here */
     if (ssl->options.dtls && ssl->dtls13WaitKeyUpdateAck)
-            return WOLFSSL_SUCCESS;
+        return 0;
```
nit: this should go in the commit that refactor `wolfSSL_send_update_key`

julek-wolfssl replied:
Changing.

rizlik commented on src/ssl.c:
```diff
@@ -3284,7 +3284,7 @@ int wolfSSL_Rehandshake(WOLFSSL* ssl)
     ssl->options.peerAuthGood = 0;
 
-#ifdef WOLFSSL_DTLS
+#ifdef WOLFSSL_DTLS13
```
why?

julek-wolfssl replied:
Not sure. Fixing.

rizlik commented on src/internal.c:
```diff
@@ -22031,6 +22034,68 @@ static int ModifyForMTU(WOLFSSL* ssl, int buffSz, int outputSz, int mtuSz)
+    }
+#ifdef WOLFSSL_DTLS13
+    if (ssl->options.dtls) {
+        Dtls13GetSeq(ssl, CUR_ORDER, seq32, 0);
```
can't we just take seq from `Dtls13EncryptedEpoch->nextSeqNumber`?

julek-wolfssl replied:
We can. Changing.

rizlik commented on wolfssl/wolfcrypt/settings.h:
```diff
@@ -2766,8 +2766,9 @@ extern void uITRON4_free(void *p) ;
-/* DTLS v1.3 requires 64-bit number wrappers */
-#if defined(WOLFSSL_DTLS13) && !defined(WOLFSSL_W64_WRAPPER)
+/* TLS v1.3 requires 64-bit number wrappers */
+#if defined(WOLFSSL_TLS13) || defined(WOLFSSL_DTLS13)
```
Are w64wrapped used as well for dtls1.2 only code?

julek-wolfssl replied:
Necessary when `WOLFSSL_DTLS_DROP_STATS` is defined.

julek-wolfssl commented on tests/api.c:
```diff
@@ -55079,47 +55118,19 @@ static int test_wolfSSL_dtls_plaintext(void)
-#ifdef WOLFSSL_TIRTOS
-        fdOpenSession(Task_self());
-#endif
+        AssertTrue(func_cb_client.return_code);
```
Fixing.

julek-wolfssl commented on src/internal.c:
```diff
@@ -80,6 +80,11 @@
+ *     Ignore the AEAD limits for messages specified in the RFC. After
+ *     reaching the limit, we initiate a key update.
+ *     https://www.rfc-editor.org/rfc/rfc8446#section-5.5
+ *     https://www.rfc-editor.org/rfc/rfc9147.html#name-aead-limits
```
I'm not sure what you want. Do you want the default value of the AEAD limit?

rizlik replied:
if `WOLFSSL_TLS13_IGNORE_AEAD_LIMITS` is defined or not by default.

julek-wolfssl replied:
It is NOT defined by default.

rizlik replied:
can you add that information to the description of the define?

julek-wolfssl commented on src/internal.c:
```diff
@@ -18771,6 +18776,98 @@ static WC_INLINE int VerifyMac(WOLFSSL* ssl, const byte* input, word32 msgSz,
+ * https://www.rfc-editor.org/rfc/rfc9147.html#name-aead-limits
+ * We specify the limit by which we need to do a key update as the halfway point
+ * to the hard decryption fail limit. */
+static int checkDTLS13AEADFailLimit(byte bulk_cipher_algorithm,
```
Moving to `dtls13.c`. I originally thought that DTLS 1.2 also defines limits but it does not.

julek-wolfssl commented on src/internal.c:
```diff
@@ -18771,6 +18776,98 @@ static WC_INLINE int VerifyMac(WOLFSSL* ssl, const byte* input, word32 msgSz,
+        WOLFSSL_MSG("Connection exceeded hard AEAD limit");
+        return AEAD_LIMIT_FAIL;
+    }
+    else if (w64GT(dropped, keyUpdateLimit)) {
```
How? We first check if we hit the hard limit and then the key update limit.

rizlik replied:
because fo the return if the main if

julek-wolfssl replied:
Can you post an example of how you would like to see this written?

julek-wolfssl commented on src/internal.c:
```diff
@@ -18771,6 +18776,98 @@ static WC_INLINE int VerifyMac(WOLFSSL* ssl, const byte* input, word32 msgSz,
+static int HandleDTLSDecryptFailed(WOLFSSL* ssl)
+{
+    ssl->options.processReply = doProcessInit;
+    ssl->buffers.inputBuffer.idx = ssl->buffers.inputBuffer.length;
```
Moving.

julek-wolfssl committed:
Code review fixes

- Mark old epochs as invalid so we don't attempt to decrypt with them
- Return a non-zero value if possible in unit tests
- Move Dtls13CheckAEADFailLimit to dtls13.c
- Reset state in processreply

julek-wolfssl committed:
Keep a separate drop counter for each epoch

rizlik reviewed (approved):
Nice work

julek-wolfssl committed:
Add prototypes for w64wrapper functions

JacobBarthelmeh commented on src/internal.c:
```diff
@@ -19077,17 +19120,14 @@ int ProcessReplyEx(WOLFSSL* ssl, int allowSocketErr)
             #endif
                 if (ret < 0) {
                     WOLFSSL_MSG("VerifyMacEnc failed");
-                    WOLFSSL_ERROR(ret);
```
this removes an error from the error queue with compat. builds, is it necessary to remove it?

JacobBarthelmeh reviewed (commented):
```
./configure --enable-dtls --enable-dtls13 CPPFLAGS=-DWOLFSSL_TLS13_IGNORE_AEAD_LIMITS
```

fails to build for me with some compile errors in tests/api.c

```
tests/api.c:55950:60: error: ‘struct Dtls13Epoch’ has no member named ‘dropCount’
     w64Decrement(&Dtls13GetEpoch(ssl, ssl->dtls13PeerEpoch)->dropCount);
                                                            ^~
  CC       tests/tests_unit_test-srp.o
```

Is there more needed to configure with this enabled?

julek-wolfssl committed:
Test AEAD limits only without WOLFSSL_TLS13_IGNORE_AEAD_LIMITS

julek-wolfssl commented:
The `api.c` tests needed to be gated on `WOLFSSL_TLS13_IGNORE_AEAD_LIMITS` too. Fixed and pushed.
